src/moq_instance.cpp

```diff
@@ -0,0 +1,1687 @@
+        }
+
+        if (not conn_ctx.ctrl_msg_type_received) { // should never happen
+            close_connection(conn_ctx.conn_id,
+                             MoQTerminationReason::INTERNAL_ERROR,
+                             "Process recv message connection context is missing message type");
+        }
+
+        switch (*conn_ctx.ctrl_msg_type_received) {
+            case MoQMessageType::SUBSCRIBE: {
+                if (not stream_buffer->anyHasValue()) {
+                    _logger->debug << "Received subscribe, init stream buffer" << std::flush;
+                    stream_buffer->initAny<MoqSubscribe>();
+                }
+
+                auto& msg = stream_buffer->getAny<MoqSubscribe>();
+                if (*stream_buffer >> msg) {
+                    auto tfn = TrackFullName{ msg.track_namespace, msg.track_name };
+                    auto th = TrackHash(tfn);
+
+                    if (msg.subscribe_id > conn_ctx._sub_id) {
+                        conn_ctx._sub_id = msg.subscribe_id + 1;
+                    }
+
+                    // For client/publisher, notify track that there is a subscriber
+                    if (_client_mode) {
+                        auto ptd = getPubTrackDelegate(conn_ctx, th);
+                        if (not ptd.has_value()) {
+                            _logger->warning
+                              << "Received subscribe unknown publish track" << " conn_id:" << conn_ctx.conn_id
+                              << " namespace hash: " << th.track_namespace_hash << " name hash: " << th.track_name_hash
+                              << std::flush;
+
+                            send_subscribe_error(conn_ctx,
+                                                 msg.subscribe_id,
+                                                 msg.track_alias,
+                                                 MoQSubscribeError::TRACK_NOT_EXIST,
+                                                 "Published track not found");
+                            return true;
+                        }
+
+                        send_subscribe_ok(conn_ctx, msg.subscribe_id, MOQT_SUBSCRIBE_EXPIRES, false);
+
+                        _logger->debug << "Received subscribe to announced track alias: " << msg.track_alias
+                                       << " recv subscribe_id: " << msg.subscribe_id
+                                       << ", setting send"
+                                          ""
+                                          " state to ready"
+                                       << std::flush;
+
+                        // Indicate send is ready upon subscribe
+                        // TODO(tievens): Maybe needs a delay as subscriber may have not received ok before data is sent
+                        ptd->lock()->setSubscribeId(msg.subscribe_id);
+                        ptd->lock()->setSendStatus(MoQTrackDelegate::TrackSendStatus::OK);
+                        ptd->lock()->cb_sendReady();
```

Comment:
Why not retrieve the lock once and store temporarily?

Reply:
Updated.